delivery_auto_refresh/models/sale_order.py

```diff
@@ -56,63 +85,109 @@ def _get_param_auto_add_delivery_line(self):
+                # Tax is set with a SET command
+                clear = update = False
+                for cmd in val:
+                    if cmd[0] == fields.Command.SET:
+                        if delivery_line[f].ids != cmd[2]:
+                            update = True
+                    else:
+                        clear = True
+                if clear:
+                    new_vals[f] = [fields.Command.CLEAR] + val
+                elif update:
+                    new_vals[f] = val
+            elif isinstance(field_def, fields.Many2one):
+                if delivery_line[f].id != val:
+                    new_vals[f] = val
+            elif f == "sequence":
+                # sequence is last sequence + 1. As the delivery line already
+                # exists, substract 1
+                if delivery_line[f] != val - 1:
+                    new_vals[f] = val
+            elif delivery_line[f] != val:
+                new_vals[f] = val
+        if new_vals:
+            delivery_line.write(new_vals)
+
     def _auto_refresh_delivery(self):
         self.ensure_one()
-        # Make sure that if you have removed the carrier, the line is gone
-        if self.state in {"draft", "sent"}:
-            # Context added to avoid the recursive calls and save the new
-            # value of carrier_id
-            self.with_context(auto_refresh_delivery=True)._remove_delivery_line()
-        if self._get_param_auto_add_delivery_line() and self.carrier_id:
-            if self.state in {"draft", "sent"}:
-                price_unit = self.carrier_id.rate_shipment(self)["price"]
-                if not self.is_all_service:
-                    self._create_delivery_line(self.carrier_id, price_unit)
-                self.with_context(auto_refresh_delivery=True).write(
-                    {"recompute_delivery_price": False}
-                )
-
-    @api.model
-    def create(self, vals):
-        """Create or refresh delivery line on create."""
-        order = super().create(vals)
-        order._auto_refresh_delivery()
-        return order
+        if self.env.context.get("auto_refresh_delivery"):
+            return
+        if not self._get_param_auto_add_delivery_line():
+            return
+        if self.state not in ("draft", "sent"):
+            return
+
+        self = self.with_context(auto_refresh_delivery=True)
+
+        if not self.carrier_id:
+            self._set_delivery_carrier()
+
+        if not self.carrier_id:
+            self._remove_delivery_line()
+        elif self.is_all_service:
+            self._remove_delivery_line()
+        else:
+            price_unit = self.carrier_id.rate_shipment(self)["price"]
+            delivery_lines = self.order_line.filtered("is_delivery")
+            if not delivery_lines:
+                self._create_delivery_line(self.carrier_id, price_unit)
+            elif len(delivery_lines) > 1:
+                delivery_discount = self._get_delivery_discount()
```

Comment:
you have already filtered the lines above. I would make this method accept the lines instead. Or trash the method unless you need it as a hook.